src/ucp/proto/proto_am.inl

```diff
@@ -83,28 +83,96 @@ err:
     return packed_len;
 }
 
-static ucs_status_t UCS_F_ALWAYS_INLINE
-ucp_do_am_zcopy_single(uct_pending_req_t *self, uint8_t am_id, const void *hdr,
-                       size_t hdr_size, ucp_req_complete_func_t complete)
+static UCS_F_ALWAYS_INLINE
+size_t ucp_dt_iov_copy_uct(uct_iov_t *iov, size_t *iovcnt, size_t max_dst_iov,
+                           ucp_frag_state_t *state, const ucp_dt_iov_t *src_iov,
+                           ucp_datatype_t datatype, size_t length_max)
+{
+    size_t iovcnt_offset, iov_offset, max_src_iov;
+    size_t src_it, src_it_offset, dst_it;
+    const uct_mem_h *memh;
+    size_t length_it = 0;
+
+    switch (datatype & UCP_DATATYPE_CLASS_MASK) {
+    case UCP_DATATYPE_CONTIG:
+        iov[0].buffer = (void *)src_iov + state->offset;
+        iov[0].length = length_max;
+        iov[0].memh   = state->dt.contig.memh;
+        iov[0].stride = 0;
+        iov[0].count  = 1;
+
+        *iovcnt   = 1;
+        length_it = iov[0].length;
+        break;
+    case UCP_DATATYPE_IOV:
+        memh          = state->dt.iov.memh;
+        iovcnt_offset = state->dt.iov.iovcnt_offset;
+        iov_offset    = state->dt.iov.iov_offset;
+        max_src_iov   = state->dt.iov.iovcnt;
+
+        for (src_it = 0, dst_it = 0; (dst_it < max_dst_iov) &&
+                                     ((src_it + iovcnt_offset) < max_src_iov) &&
+                                     !(length_it >= length_max); ++src_it) {
+            src_it_offset = src_it + iovcnt_offset;
```

Comment:
looks like we don't really need both src_it and src_it_offser variables, and use just 'src_it' and instead of starting it from 0, start it from iovcnt_offset.